Unable to install or uninstall plugins when plugins pre-installed at Docker image creation time using ETHERPAD_PLUGINS argument

I am using the Docker install of Etherpad by setting the ETHERPAD_PLUGINS argument in a docker-compose file (this would be the same as if I set them in the Dockerfile).

This is correctly installing the specified plugins into the created image. However, no new plugins can be installed while trying to use the admin interface. And the existing installed plugins cannot be uninstalled without breaking the whole installation.

The error I see when trying to install a new plugin is:
npm WARN enoent ENOENT: no such file or directory, open '/opt/etherpad-lite/package.json'.
However the plugin folder under node_modules has been created in the container but doesn't appear as installed in the admin interface.

When I try and remove any plugin the error I get is the following and the whole installation crashes and cannot be started up again. This can be fixed by deleting the package-lock.json in the container.
Error when removing plugin (ep_align in this case. But the same happens with any plugin):

```
[2021-06-14 18:30:10.752] [ERROR] runCmd|npm - npm WARN enoent ENOENT: no such file or directory, open '/opt/etherpad-lite/package.json'
[2021-06-14 18:30:10.762] [ERROR] runCmd|npm - npm WARN etherpad-lite No description
[2021-06-14 18:30:10.770] [ERROR] runCmd|npm - npm WARN etherpad-lite No repository field.
[2021-06-14 18:30:10.778] [ERROR] runCmd|npm - npm WARN etherpad-lite No README data
[2021-06-14 18:30:10.787] [ERROR] runCmd|npm - npm WARN etherpad-lite No license field.
[2021-06-14 18:30:10.788] [ERROR] runCmd|npm -
[2021-06-14 18:30:16.141] [ERROR] runCmd|npm - npm ERR! missing: ep_align@0.3.34, required by etherpad-lite
[2021-06-14 18:30:16.159] [ERROR] server - Metrics at time of fatal error:
{
  "httpStartTime": 1623694990073,
  "memoryUsage": 112902144,
  "memoryUsageHeap": 40922848,
  "ueberdb_lockAwaits": 0,
  "ueberdb_lockAcquires": 167,
  "ueberdb_lockReleases": 167,
  "ueberdb_reads": 163,
  "ueberdb_readsFailed": 0,
  "ueberdb_readsFinished": 163,
  "ueberdb_readsFromCache": 143,
  "ueberdb_readsFromDb": 20,
  "ueberdb_readsFromDbFailed": 0,
  "ueberdb_readsFromDbFinished": 20,
  "ueberdb_writes": 9,
  "ueberdb_writesFailed": 0,
  "ueberdb_writesFinished": 9,
  "ueberdb_writesObsoleted": 0,
  "ueberdb_writesToDb": 9,
  "ueberdb_writesToDbFailed": 0,
  "ueberdb_writesToDbFinished": 9,
  "totalUsers": 5,
  "connects": {
    "mean": 0.02285118625882035,
    "count": 18,
    "currentRate": 0.022851184693131816,
    "1MinuteRate": 0.012563832192270178,
    "5MinuteRate": 0.012201679444538326,
    "15MinuteRate": 0.01110730824014481
  },
  "activePads": 0,
  "httpRequests": {
    "meter": {
      "mean": 0.02424904251600877,
      "count": 19,
      "currentRate": 0.024249042567011696,
      "1MinuteRate": 0.026038478006953157,
      "5MinuteRate": 0.013952997839964863,
      "15MinuteRate": 0.011706946222158026
    },
    "histogram": {
      "min": 2.0739920139312744,
      "max": 1001.4482519924641,
      "sum": 1116.1726340949535,
      "variance": 52145.67777190968,
      "mean": 58.745928110260714,
      "stddev": 228.3542812646824,
      "count": 19,
      "median": 4.190887004137039,
      "p75": 10.734652996063232,
      "p95": 1001.4482519924641,
      "p99": 1001.4482519924641,
      "p999": 1001.4482519924641
    }
  },
  "lastDisconnect": 1623695416159,
  "disconnects": {
    "mean": 0.018390921739918315,
    "count": 13,
    "currentRate": 0.018390921827336034,
    "1MinuteRate": 0.012552036501397008,
    "5MinuteRate": 0.010323871975289348,
    "15MinuteRate": 0.008472578169404025
  }
}
[2021-06-14 18:30:16.161] [ERROR] server - Error: Command exited with code 1: npm ls --long --json --depth=0 --no-production
    at exports (/opt/etherpad-lite/src/node/utils/run_cmd.js:119:25)
    at Object.exports.getPackages (/opt/etherpad-lite/src/static/js/pluginfw/plugins.js:125:48)
    at Object.exports.update (/opt/etherpad-lite/src/static/js/pluginfw/plugins.js:99:34)
    at Object.exports.uninstall (/opt/etherpad-lite/src/static/js/pluginfw/installer.js:45:17)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
vfa-etherpad exited with code 1
```

Error when trying to start Etherpad back up after crash:

```
[2021-06-14 18:33:25.361] [ERROR] runCmd|npm - npm ERR! missing: ep_align@0.3.34, required by etherpad-lite
[2021-06-14 18:33:25.377] [ERROR] server - Error occurred while starting Etherpad
[2021-06-14 18:33:25.378] [ERROR] server - Metrics at time of fatal error:
{
  "httpStartTime": 0,
  "memoryUsage": 61988864,
  "memoryUsageHeap": 13371584,
  "ueberdb_lockAwaits": 0,
  "ueberdb_lockAcquires": 0,
  "ueberdb_lockReleases": 0,
  "ueberdb_reads": 0,
  "ueberdb_readsFailed": 0,
  "ueberdb_readsFinished": 0,
  "ueberdb_readsFromCache": 0,
  "ueberdb_readsFromDb": 0,
  "ueberdb_readsFromDbFailed": 0,
  "ueberdb_readsFromDbFinished": 0,
  "ueberdb_writes": 0,
  "ueberdb_writesFailed": 0,
  "ueberdb_writesFinished": 0,
  "ueberdb_writesObsoleted": 0,
  "ueberdb_writesToDb": 0,
  "ueberdb_writesToDbFailed": 0,
  "ueberdb_writesToDbFinished": 0
}
[2021-06-14 18:33:25.378] [ERROR] server - Error: Command exited with code 1: npm ls --long --json --depth=0 --no-production
    at exports (/opt/etherpad-lite/src/node/utils/run_cmd.js:119:25)
    at Object.exports.getPackages (/opt/etherpad-lite/src/static/js/pluginfw/plugins.js:125:48)
    at Object.exports.update (/opt/etherpad-lite/src/static/js/pluginfw/plugins.js:99:34)
    at Object.exports.start (/opt/etherpad-lite/src/node/server.js:139:19)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
[2021-06-14 18:33:30.378] [ERROR] server - Something that should have been cleaned up during the shutdown hook (such as a timer, worker thread, or open connection) is preventing Node.js from exiting
[2021-06-14 18:33:30.380] [ERROR] server - Forcing an unclean exit...
```

If no plugins are installed at the time of the Docker image creation by not specifying any plugins in the ETHERPAD_PLUGINS argument then it all works as expected.


**To Reproduce**
1. Create etherpad docker image with specifying ETHERPAD_PLUGINS argument
2. Run docker image
3. Using admin interface try to add any plugin or remove any plugin
4. For adding a plugin it errors. For removing a plugin it crashes and won't start up again unless the package-lock.json is deleted from the container.

**Expected behavior**
Plugins should be able to be installed and uninstalled even if some are already installed as part of the Docker image creation.



**Server**
 - Etherpad version: 1.8.13
 - OS: Ubuntu 18.04
 - Docker: Docker version 20.10.7, build f0df350
 - node version (inside container): v14.17.0
 - npm version (`npm --version`): 6.14.13

**Desktop**
 - OS: MacOS Big Sur
 - Browser: Chrome
 - Version: 91.0.4472.101 (Official Build) (x86_64)


**UPDATE**
There is a package-lock.json file created when the image is created with ETHERPAD_PLUGINS argument set. If this file is deleted manually from the container then installing and uninstalling plugins starts to work from the admin interface.

--- a/Dockerfile
+++ b/Dockerfile
@@ -55,12 +55,17 @@ RUN groupadd --system ${EP_GID:+--gid "${EP_GID}" --non-unique} etherpad && \
 ARG EP_DIR=/opt/etherpad-lite
 RUN mkdir -p "${EP_DIR}" && chown etherpad:etherpad "${EP_DIR}"
 
-# install abiword for DOC/PDF/ODT export
-RUN [ -z "${INSTALL_ABIWORD}" ] || (apt update && apt -y install abiword && apt clean && rm -rf /var/lib/apt/lists/*)
-
-# install libreoffice for DOC/PDF/ODT export
-# the mkdir is needed for configuration of openjdk-11-jre-headless, see https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=863199
-RUN [ -z "${INSTALL_SOFFICE}" ] || (apt update && mkdir -p /usr/share/man/man1 && apt -y install libreoffice && apt clean && rm -rf /var/lib/apt/lists/*)
+# the mkdir is needed for configuration of openjdk-11-jre-headless, see
+# https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=863199
+RUN [ -n "${INSTALL_ABIWORD}${INSTALL_SOFFICE}" ] || exit 0; \
+    mkdir -p /usr/share/man/man1 && \
+    apt update && \
+    apt -y install \
+        ${INSTALL_ABIWORD:+abiword} \
+        ${INSTALL_SOFFICE:+libreoffice} \
+        && \
+    apt clean && \
+    rm -rf /var/lib/apt/lists/*
 
 USER etherpad
 

--- a/Dockerfile
+++ b/Dockerfile
@@ -61,7 +61,7 @@ RUN [ -n "${INSTALL_ABIWORD}${INSTALL_SOFFICE}" ] || exit 0; \
     export DEBIAN_FRONTEND=noninteractive; \
     mkdir -p /usr/share/man/man1 && \
     apt-get -qq update && \
-    apt-get -qq install \
+    apt-get -qq --no-install-recommends install \
         ${INSTALL_ABIWORD:+abiword} \
         ${INSTALL_SOFFICE:+libreoffice} \
         && \

--- a/Dockerfile
+++ b/Dockerfile
@@ -59,12 +59,12 @@ RUN mkdir -p "${EP_DIR}" && chown etherpad:etherpad "${EP_DIR}"
 # https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=863199
 RUN [ -n "${INSTALL_ABIWORD}${INSTALL_SOFFICE}" ] || exit 0; \
     mkdir -p /usr/share/man/man1 && \
-    apt update && \
-    apt -y install \
+    apt-get -qq update && \
+    apt-get -qq install \
         ${INSTALL_ABIWORD:+abiword} \
         ${INSTALL_SOFFICE:+libreoffice} \
         && \
-    apt clean && \
+    apt-get -qq clean && \
     rm -rf /var/lib/apt/lists/*
 
 USER etherpad

--- a/Dockerfile
+++ b/Dockerfile
@@ -58,6 +58,7 @@ RUN mkdir -p "${EP_DIR}" && chown etherpad:etherpad "${EP_DIR}"
 # the mkdir is needed for configuration of openjdk-11-jre-headless, see
 # https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=863199
 RUN [ -n "${INSTALL_ABIWORD}${INSTALL_SOFFICE}" ] || exit 0; \
+    export DEBIAN_FRONTEND=noninteractive; \
     mkdir -p /usr/share/man/man1 && \
     apt-get -qq update && \
     apt-get -qq install \

--- a/Dockerfile
+++ b/Dockerfile
@@ -68,11 +68,10 @@ WORKDIR "${EP_DIR}"
 
 COPY --chown=etherpad:etherpad ./ ./
 
-# install node dependencies for Etherpad
 RUN src/bin/installDeps.sh && \
-	rm -rf ~/.npm/_cacache
-
-RUN [ -z "${ETHERPAD_PLUGINS}" ] || npm install --no-save ${ETHERPAD_PLUGINS}
+    { [ -z "${ETHERPAD_PLUGINS}" ] || \
+      npm install --no-save ${ETHERPAD_PLUGINS}; } && \
+    rm -rf ~/.npm
 
 # Copy the configuration file.
 COPY --chown=etherpad:etherpad ./settings.json.docker "${EP_DIR}"/settings.json

--- a/Dockerfile
+++ b/Dockerfile
@@ -57,11 +57,12 @@ RUN mkdir -p "${EP_DIR}" && chown etherpad:etherpad "${EP_DIR}"
 
 # the mkdir is needed for configuration of openjdk-11-jre-headless, see
 # https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=863199
-RUN [ -n "${INSTALL_ABIWORD}${INSTALL_SOFFICE}" ] || exit 0; \
-    export DEBIAN_FRONTEND=noninteractive; \
+RUN export DEBIAN_FRONTEND=noninteractive; \
     mkdir -p /usr/share/man/man1 && \
     apt-get -qq update && \
     apt-get -qq --no-install-recommends install \
+        ca-certificates \
+        git \
         ${INSTALL_ABIWORD:+abiword} \
         ${INSTALL_SOFFICE:+libreoffice} \
         && \

--- a/Dockerfile
+++ b/Dockerfile
@@ -72,7 +72,7 @@ COPY --chown=etherpad:etherpad ./ ./
 RUN src/bin/installDeps.sh && \
 	rm -rf ~/.npm/_cacache
 
-RUN [ -z "${ETHERPAD_PLUGINS}" ] || npm install ${ETHERPAD_PLUGINS}
+RUN [ -z "${ETHERPAD_PLUGINS}" ] || npm install --no-save ${ETHERPAD_PLUGINS}
 
 # Copy the configuration file.
 COPY --chown=etherpad:etherpad ./settings.json.docker "${EP_DIR}"/settings.json

--- a/Dockerfile
+++ b/Dockerfile
@@ -68,9 +68,17 @@ WORKDIR "${EP_DIR}"
 
 COPY --chown=etherpad:etherpad ./ ./
 
-RUN src/bin/installDeps.sh && \
-    { [ -z "${ETHERPAD_PLUGINS}" ] || \
+# Plugins must be installed before installing Etherpad's dependencies, otherwise
+# npm will try to hoist common dependencies by removing them from
+# src/node_modules and installing them in the top-level node_modules. As of
+# v6.14.10, npm's hoist logic appears to be buggy, because it sometimes removes
+# dependencies from src/node_modules but fails to add them to the top-level
+# node_modules. Even if npm correctly hoists the dependencies, the hoisting
+# seems to confuse tools such as `npm outdated`, `npm update`, and some ESLint
+# rules.
+RUN { [ -z "${ETHERPAD_PLUGINS}" ] || \
       npm install --no-save ${ETHERPAD_PLUGINS}; } && \
+    src/bin/installDeps.sh && \
     rm -rf ~/.npm
 
 # Copy the configuration file.

